src/pudl/transform/ferc1.py

```diff
@@ -97,6 +101,138 @@ class Ferc1RenameColumns(TransformParams):
     xbrl: RenameColumns = {}
 
 
+class WideToTidyXBRL(TransformParams):
+    """Parameters for converting a wide XBRL table to a tidy table with value types."""
+
+    idx_cols: list[str] | None
+    """List of column names to treat as the table index."""
+
+    value_types: list[str] | None
+    """List of names of value types.
+
+    These will be the renamed XBRL column suffixes and will be the output table columns."""
+
+    expected_drop_cols: int = 0
+    """The number of columns that are expected to be dropped.
+
+    :func:`wide_to_tidy_xbrl` will generate a regex pattern assuming the ``value_types``
+    are the column name's suffixes. If a column does not conform to that pattern, it
+    will be filtered out. This is helpful for us to not include a bunch of columns from
+    the input dataframe incorrectly included in the stacking process. We could enumerate
+    every column that we want to drop, but this could be tedious and potentially error
+    prone. But this does mean that if a column is incorrectly named - or barely missing
+    the pattern, it will be dropped. This parameter enables us to lock the number of
+    expected columns. If the dropped columns are a different number, an error will be
+    raised.
+    """
+
+
+def wide_to_tidy_xbrl(df: pd.DataFrame, params: WideToTidyXBRL) -> pd.DataFrame:
+    """Reshape wide tables with FERC account columns to tidy format.
+
+    The XBRL table coming into this method contains all the data from both the
+    instant and duration tables in a wide format -- with one column for every
+    combination of value type (e.g. additions, ending_balance) and accounting
+    category, which means ~500 columns.
+
+    We tidy this into a long table with one column for each of the value types (6 in
+    all), and a new column that contains the accounting categories. This allows
+    aggregation across columns to calculate the ending balance based on the starting
+    balance and all of the reported changes, and aggregation across groups of rows
+    to total up various hierarchical accounting categories (hydraulic turbines ->
+    hydraulic production plant -> all production plant -> all electric utility
+    plant) though the categorical columns required for that aggregation are added
+    later.
+    """
+    suffixes = "|".join(params.value_types)
+    pat = r"(^.*)_(" + suffixes + r"$)"
+    # filter out any columns that don't match the pattern
+    df_out = df.set_index(params.idx_cols).filter(regex=pat)
+    # check if there are unexpected columns being dropped
+    dropped_cols = [col for col in df if col not in df_out.reset_index()]
```

Comment:
I guess it's not strictly necessary but I think it's clearer what's going on if we refer to the columns specifically, rather than iterating over the naked dataframe, since there are multiple ways one could imagine iterating over a dataframe:
```py
dropped_cols = [col for col in df.columns if col not in df_out.reset_index().columns]
```

Reply:
sure. the default behavior when iterating like this is to iterate over the columns. but it is more explicit this way